### PR TITLE
nixos-container: fix allow alternative nixos paths

### PR DESCRIPTION
**Pull request by ericbmerritt**
###### Things done
- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [x] NixOS
  - [ ] OS X
  - [ ] Linux
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

This should be completely backwards compatible. It allows the '-f' part
of the nix-env command to be configured. This greatly eases using
nixos-container as part of development where several nixpkgs
repositories might be tested at the same time.

---

This change is [Reviewable](https://reviewable.io/reviews/nixos/nixpkgs/15967)

### CODE REVIEW

**Comment by joachifm** [comment]
cc @edolstra

**Comment by lethalman** [comment]
@edolstra looks fine, perhaps it should also be supported for declarative containers? I guess it's not necessary, right? Or do you see potential problems?

**Comment by ericbmerritt** [comment]
@lethalman You are probably correct. I use it exclusively for on the fly containers, but I could see good reasons for having it in the declarative containers as well.

**Comment by lethalman** [comment]
@ericbmerritt I'm ok with the change, I just wonder if it may break declarative containers at first sight...

**Comment by ericbmerritt** [comment]
In the case where no path is passed, it should have the exact same behaviour as before.

**Comment by ericbmerritt** [comment]
@lethalman I just added an additional change that makes containers quite a bit easier to use for testing configurations. Both of these should be backwards compatible.

File: nixos/modules/virtualisation/nixos-container.pl [hunk 1]
```diff
@@ -68,6 +73,15 @@ sub showHelp {
 sub writeNixOSConfig {
     my ($nixosConfigFile) = @_;
 
+    my $localExtraConfig = "";
+
+    if ($extraConfig) {
+        $localExtraConfig = $extraConfig
+    } elsif ($configFile) {
```
Comment (fpletz) [inline nixos/modules/virtualisation/nixos-container.pl]: I see a potential problem here. If both `--config` and `--config-file` are provided, only the `--config` parameter will be respected and `--config-file` will be ignored silently, right? We should probably either merge `$localExtraConfig` or exit with an error.

**Comment by fpletz** [comment]
Apart from my comment, :+1: from me.

**Comment by ericbmerritt** [comment]
@fpietz updated

**Comment by ericbmerritt** [comment]
is there something more I need to do here?

File: nixos/modules/virtualisation/nixos-container.pl [hunk 2]
```diff
@@ -34,20 +35,28 @@ sub showHelp {
...
+    "nixos-path=s" => \$nixosPath,
+    "config=s" => \$extraConfig,
+    "config-file=s" => \$configFile
     ) or exit 1;
 
 my $action = $ARGV[0] or die "$0: no action specified\n";
 
+if (defined $configFile and defined $extraConfig) {
+    die "--config and --config-file are mutually incompatible. " .
+        "Please define on or the other, but not both";
```
Comment (danbst) [inline nixos/modules/virtualisation/nixos-container.pl]: typo "on"

**Comment by joachifm** [comment]
@ericbmerritt looks like the consensus is yay on this, so if you rebase, I'll merge it.

**Comment by ericbmerritt** [comment]
@joachifm  rebased.

**Comment by edolstra** [comment]
I don't understand the need for this flag. Why not just set $NIX_PATH to point to the alternative NixOS tree?

**Comment by ericbmerritt** [comment]
@edolstra nix_path is ignored by the underlying command.

**Comment by cleverca22** [comment]
nix-env does obey the search path when it encounters a <search> but by default, it loads directly from the channel and never tries to search for things

**Pull request merged**